### PR TITLE
Make deprecated reason non-null

### LINKED ISSUE #3736: Tracking spec change: non-nullable reason for `@deprecated`
Keeping a note here for us to return to later, this small spec change may get merged in soon https://github.com/graphql/graphql-spec/pull/1040

It changes `@deprecated`'s reason to be non-nullable. There's a default value for the reason already, so the migration pathway isn't a big deal, it's more about removing an odd edge case.

This is not yet in the spec, this issue is a reminder to check in on this topic again later

### PR DESCRIPTION
**Pull request by dondonz**
Closes #3736

An incoming spec change will require the `@deprecated` reason to be non-nullable. https://github.com/graphql/graphql-spec/pull/1040. 

Whilst this is technically a breaking change, the reason already has a default value, so it is only a breaking change if the reason `null` was previously used.

Update: when I started the PR it was at RFC1 stage, at the time of merging the proposal reached RFC3 (Accepted) stage and is already in the draft spec document: https://spec.graphql.org/draft/#sec--deprecated

### CODE REVIEW

File: src/main/java/graphql/util/Anonymizer.java [hunk 1]
```diff
@@ -260,16 +259,6 @@ public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition graph
...
-                if (Directives.DEPRECATED_DIRECTIVE_DEFINITION.getName().equals(graphQLDirective.getName())) {
-                    GraphQLArgument reason = newArgument().name("reason")
-                            .type(Scalars.GraphQLString)
-                            .clearValue().build();
-                    GraphQLDirective newElement = graphQLDirective.transform(builder -> {
-                        builder.description(null).argument(reason);
-                    });
-                    changeNode(context, newElement);
-                    return TraversalControl.ABORT;
-                }
```
Comment (dondonz) [inline src/main/java/graphql/util/Anonymizer.java:272]: I am guessing the Anonymizer pre-dates the split of GraphQLDirective and GraphQLAppliedDirective

Now that applied directives are separated, there shouldn't be any transformation in this `GraphQLDirective` visitor method. The deprecated directive as defined in the schema should be unchanged, as any other built-in directive in the next lines

```
if (DirectiveInfo.isGraphqlSpecifiedDirective(graphQLDirective.getName())) {
    return TraversalControl.ABORT;
}
```

File: src/main/java/graphql/util/Anonymizer.java [hunk 2]
```diff
@@ -285,7 +274,8 @@ public TraversalControl visitGraphQLAppliedDirective(GraphQLAppliedDirective gra
                 if (Directives.DEPRECATED_DIRECTIVE_DEFINITION.getName().equals(graphQLDirective.getName())) {
```
Comment (dondonz) [inline src/main/java/graphql/util/Anonymizer.java:274]: As I mentioned above, the censoring of `@deprecated` reasons should actually only be here, in the `GraphQLAppliedDirective` visitor method only

File: src/main/java/graphql/util/Anonymizer.java [hunk 3]
```diff
@@ -285,7 +274,8 @@ public TraversalControl visitGraphQLAppliedDirective(GraphQLAppliedDirective gra
...
                     GraphQLAppliedDirectiveArgument reason = GraphQLAppliedDirectiveArgument.newArgument().name("reason")
                             .type(Scalars.GraphQLString)
```
Comment (dondonz) [inline src/main/java/graphql/util/Anonymizer.java:276]: It's ok for an applied directive's argument to be nullable. In this case, the default value ("No longer supported") as defined in the schema will be used

File: src/test/groovy/graphql/schema/idl/SchemaPrinterTest.groovy [hunk 4]
```diff
@@ -968,7 +968,7 @@ type Query {
 "Marks the field, argument, input field or enum value as deprecated"
 directive @deprecated(
     "The reason for the deprecation"
-    reason: String = "No longer supported"
```
Comment (dondonz) [inline src/test/groovy/graphql/schema/idl/SchemaPrinterTest.groovy:971]: All that follows is a number of tests which need updating

File: src/main/java/graphql/Directives.java [hunk 5]
```diff
@@ -63,7 +63,7 @@ public class Directives {
                         newInputValueDefinition()
                                 .name("reason")
                                 .description(createDescription("The reason for the deprecation"))
-                                .type(newTypeName().name(STRING).build())
+                                .type(newNonNullType(newTypeName().name(STRING).build()).build())
```
Comment (bbakerman) [inline src/main/java/graphql/Directives.java:66]: what is `newNonNullType` ???

Comment (bbakerman) [reply src/main/java/graphql/Directives.java:66]: ahh a builder

**Review by bbakerman — APPROVED** [review]

**Comment by dondonz** [comment]
This proposed change has been included in the draft specification today: https://spec.graphql.org/draft/#sec--deprecated

https://github.com/graphql/graphql-spec/pull/1040

**Pull request merged**